src/components/TagManagement/AddTagForm.tsx

```diff
@@ -0,0 +1,231 @@
+import * as React from 'react';
+import styled from 'styled-components';
+import * as uuid from 'uuid';
+
+import { ResourceTagInfo } from './models';
+import some from 'lodash/some';
+import find from 'lodash/find';
+import startsWith from 'lodash/startsWith';
+import isEmpty from 'lodash/isEmpty';
+import { Alert as AlertBase } from '../Alert';
+import { Txt } from '../Txt';
+import { Input } from '../Input';
+import { Button } from '../Button';
+import {
+	SimpleConfirmationModal,
+	SimpleConfirmationModalProps,
+} from '../../internal/SimpleConfirmationModal';
+import { withPreventDefault, stopKeyDownEvent } from '../../utils';
+import { useTranslation } from '../../hooks/useTranslation';
+
+const Alert = styled(AlertBase)`
+	padding: 8px 0 18px;
+`;
+
+const NewTagTr = styled.tr`
+	table > thead > & > td {
+		padding-top: 0;
+		padding-bottom: 0;
+	}
+`;
+
+const NewTagErrorsTr = styled.tr`
+	table > thead > & > td {
+		height: 30px;
+		padding-top: 5px;
+		padding-bottom: 5px;
+	}
+`;
+
+const RESERVED_NAMESPACE = 'io.resin.';
+
+interface AddTagFormProps<T> {
+	t: any;
+	itemType: string;
+	existingTags: Array<ResourceTagInfo<T>>;
+	overwritableTags?: Array<ResourceTagInfo<T>>;
+	addTag: (tag: ResourceTagInfo<T>) => void;
+}
+
+export const AddTagForm = <T extends {}>({
+	itemType,
+	existingTags,
+	overwritableTags,
+	addTag,
+}: AddTagFormProps<T>) => {
+	const { t } = useTranslation();
+	const [tagKey, setTagKey] = React.useState<string>();
+	const [value, setValue] = React.useState<string>();
+	const [tagKeyIsInvalid, setTagKeyIsInvalid] = React.useState(false);
+	const [error, setError] = React.useState<{ message: string }>();
+	const [canSubmit, setCanSubmit] = React.useState(false);
+	const [
+		confirmationModalOptions,
+		setConfirmationModalOptions,
+	] = React.useState<SimpleConfirmationModalProps>();
+
+	const tagKeyInput = React.useRef<HTMLInputElement>(null);
+	const valueInput = React.useRef<HTMLInputElement>(null);
+	const formUuid: string = `add-tag-form-${uuid.v1()}`;
+
+	const newTagValidationRules = (key: string) => {
+		return [
+			{
+				test: () => !key || isEmpty(key),
+				message: t('fields_errors.tag_name_cannot_be_empty'),
+			},
+			{
+				test: () => /\s/.test(key!),
+				message: t('fields_errors.tag_names_cannot_contain_whitespace'),
+			},
+			{
+				test: () => startsWith(key, RESERVED_NAMESPACE),
+				message: t(`fields_errors.some_tag_keys_are_reserved`, {
+					namespace: RESERVED_NAMESPACE,
+				}),
+			},
+			{
+				test: () =>
+					some(
+						existingTags,
+						(tag) => tag.state !== 'deleted' && tag.tag_key === key,
+					),
+				message: t('fields_errors.tag_with_same_name_exists'),
+			},
+		];
+	};
```

Comment:
Let's rewrite like to accept a `t` arg and move it outside the component so that we don't re-defined it on each re-render.
```suggestion
	const newTagValidationRules = (t: ReturnType<typeof useTranslation>['t'], key: string) => {
		return [
			{
				test: () => !key || isEmpty(key),
				message: t('fields_errors.tag_name_cannot_be_empty'),
			},
			{
				test: () => /\s/.test(key!),
				message: t('fields_errors.tag_names_cannot_contain_whitespace'),
			},
			{
				test: () => startsWith(key, RESERVED_NAMESPACE),
				message: t(`fields_errors.some_tag_keys_are_reserved`, {
					namespace: RESERVED_NAMESPACE,
				}),
			},
			{
				test: () =>
					some(
						existingTags,
						(tag) => tag.state !== 'deleted' && tag.tag_key === key,
					),
				message: t('fields_errors.tag_with_same_name_exists'),
			},
		];
	};
```